### PR TITLE
feat(GSDT 424): Implement written response task feature

### PR DESCRIPTION
**Pull request by Proc3ssa**
**Overview**
This PR introduces a new Written Response Task feature that allows users to complete writing tasks with text formatting capabilities and voice-to-text transcription support.

**Key Features**
Written Response Task: Complete writing tasks with countdown timer and progress tracking
Text Formatting: Bold, italic, underline, strikethrough, and code formatting
Voice Input: Real-time speech-to-text using Web Speech API
Auto-save: User input automatically saved to store
Review Mode: Submit and review responses before finalizing

**Testing**
Added 53 comprehensive unit tests (25 for text-area, 28 for written-response)
All tests passing with 100% type safety (zero any types)
Full coverage of component lifecycle, user interactions, and edge cases

**Technical Details**
Proper TypeScript types for Speech Recognition API
OnPush change detection for performance
NgRx integration for state management
Cleanup handled in ngOnDestroy

 **Checklist**
- [ ]  All tests passing
- [ ]  No any types
- [ ]  Proper error handling
- [ ]  Memory leaks prevented

download

### CODE REVIEW

File: src/app/features/written-response/components/text-area/text-area.html [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      <textarea
+        #textarea
+        [(ngModel)]="text"
+        (input)="onTextChange()"
+        (keydown)="$event.stopPropagation();" 
+        placeholder="This is a message box"
+        class="text-input"
+      ></textarea>
+      
+      @if (text.length > 0) {
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.html]:
```suggestion
      @if (text.length) {
```

File: src/app/features/written-response/components/text-area/text-area.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+  text: string = '';
+  isRecording: boolean = false;
+  @Output() textChange = new EventEmitter<string>();
+
+  private recognition: SpeechRecognition | null = null;
+  private isRecognitionSupported: boolean = false;
+
+  constructor(private cdr: ChangeDetectorRef) {
+    // Check if browser supports Speech Recognition
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.ts]: Remove comments

File: src/app/features/written-response/components/text-area/text-area.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    const SpeechRecognitionClass = window.SpeechRecognition || window.webkitSpeechRecognition;
+    
+    if (SpeechRecognitionClass) {
+      this.isRecognitionSupported = true;
+      this.recognition = new SpeechRecognitionClass();
+      this.recognition.continuous = true;
+      this.recognition.interimResults = true;
+      this.recognition.lang = 'en-US';
+
+      // Handle speech recognition results
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.ts]: Remove comments

File: src/app/features/written-response/components/text-area/text-area.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        if (finalTranscript) {
+          this.text += finalTranscript;
+          this.onTextChange();
+          this.cdr.markForCheck();
+        }
+      };
+
+      // Handle errors
+      this.recognition.onerror = (event: SpeechRecognitionErrorEvent) => {
+        console.error('Speech recognition error:', event.error);
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.ts]: Remove log and statement

File: src/app/features/written-response/components/text-area/text-area.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        }
+
+        if (finalTranscript) {
+          this.text += finalTranscript;
+          this.onTextChange();
+          this.cdr.markForCheck();
+        }
+      };
+
+      // Handle errors
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.ts]: Remove comments

File: src/app/features/written-response/components/text-area/text-area.ts
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.ts:1]: Remove all comments and log statements

File: src/app/features/written-response/components/text-area/text-area.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    }
+  }
+  
+  onTextChange(): void {
+    this.textChange.emit(this.text);
+  }
+
+  toggleRecording(): void {
+    if (!this.isRecognitionSupported || !this.recognition) {
+      alert('Speech recognition is not supported in your browser. Please use Chrome, Edge, or Safari.');
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/components/text-area/text-area.ts]: Is alert the way team has agreed to show messages to users?

File: src/app/features/written-response/store/written-response.effects.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    private writtenResponseService: WrittenResponseService,
+    private toastService: ToastService,
+  ) {}
+
+  public loadWrittenResponseTask$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(WrittenResponseActions.loadWrittenResponseTask),
+
+      switchMap((action) =>
+        this.writtenResponseService.fetchWrittenResponseTask(action.taskId).pipe(
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response.effects.ts]: let us destructure here

File: src/app/features/written-response/store/written-response.effects.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+          map((response) =>
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response.effects.ts]: let us destructure here

File: src/app/features/written-response/store/written-response.effects.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            WrittenResponseActions.loadWrittenResponseTaskSuccess({
+              task: response.data,
+            }),
+          ),
+
+          catchError((error) => {
+            const errorMessage =
+              error.error?.message ||
+              error.message ||
+              'An unknown error occurred while loading the written response task.';
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response.effects.ts]: this keeps on repeating in the code... use a method to handle this.

File: src/app/features/written-response/store/written-response.selectors.ts
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response.selectors.ts:1]: Destructure the state properties for a cleaner look

File: src/app/features/written-response/store/written-response-service.ts
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response-service.ts:1]: Add unit test for this spec file

File: src/app/features/written-response/written-response.html [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            <p>
+              Hint: <br />
+              @if ((hints$ | async)?.length) {
+                {{ (hints$ | async)![0] }}
+              } @else {
+                No hints provided for this task.
+              }
+            </p>
+          </div>
+        </div>
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.html:78]: Why use the async pipe twice for the a similar set of information.... find a better way to handle this

File: src/app/features/written-response/written-response.spec.ts [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+    mockActivatedRoute = {
+      snapshot: {
+        paramMap: {
+          get: jest.fn().mockReturnValue('task-123'),
+        },
+      } as unknown as ActivatedRoute['snapshot'],
+    };
+
+    // Setup store selectors
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.spec.ts]: Remove comments

File: src/app/features/written-response/written-response.ts [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+  public userAnswer$: Observable<string> = this.store.select(selectWrittenResponseUserAnswer);
+  public loading$: Observable<boolean> = this.store.select(selectWrittenResponseLoading);
+  public error$: Observable<string | null> = this.store.select(selectWrittenResponseError);
+  public expectedDuration$: Observable<number> = this.store.select(
+    selectWrittenResponseExpectedDuration,
+  );
+  public isSubmitting$: Observable<boolean> = this.store.select(selectWrittenResponseIsSubmitting);
+  public submissionStatus$: Observable<string | null> = this.store.select(
+    selectWrittenResponseSubmissionStatus,
+  );
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: make use of selectSignals instead... this will reduce the rate of  usage of the async pipe in the template

File: src/app/features/written-response/written-response.ts [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+  constructor(
+    private cd: ChangeDetectorRef,
+    private store: Store,
+    private route: ActivatedRoute,
+    private toast: ToastService,
+    private router: Router,
+  ) {}
+
+  ngOnInit(): void {
+    this.taskId = this.route.snapshot.paramMap.get('id');
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: make use  of the ngrx selector we are using instead of this approach

File: src/app/features/written-response/written-response.ts [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+    if (this.taskId) {
+      this.store.dispatch(WrittenResponseActions.loadWrittenResponseTask({ taskId: this.taskId }));
+    } else {
+      this.toast.showError('Task Error', 'The task is not available, try a different task');
+      this.router.navigateByUrl('/dashboard/tasks');
+      this.store.dispatch(
+        WrittenResponseActions.loadWrittenResponseTaskFailure({ error: 'Task ID not provided.' }),
+      );
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: routing to another page and later dispatching an action afterwards, are you sure this action will be triggered ?

File: src/app/features/written-response/written-response.ts [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    this.submissionStatusSubscription = this.submissionStatus$.subscribe((status) => {
+      if (status === 'PENDING') {
+        this.progressValue = 100;
+        this.completeQuiz();
+        this.cd.markForCheck();
+      }
+    });
+
+    this.progressValue = 0;
+  }
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: Unsubscribe to after using this

**Review by Kwabena-Owusu-Achiaw — CHANGES_REQUESTED** [review]

File: src/app/features/written-response/store/written-response.effects.ts [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+  ) {}
+
+  public loadWrittenResponseTask$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(WrittenResponseActions.loadWrittenResponseTask),
+      switchMap((action) =>
+        this.writtenResponseService.fetchWrittenResponseTask(action.taskId).pipe(
+          map((response) =>
+            WrittenResponseActions.loadWrittenResponseTaskSuccess({
+              task: response.data,
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response.effects.ts]: Destructure data from response

File: src/app/features/written-response/written-response.spec.ts [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+  let mockUserAnswer: WritableSignal<string>;
+  let mockLoading: WritableSignal<boolean>;
+  let mockError: WritableSignal<string | null>;
+  let mockExpectedDuration: WritableSignal<number>;
+  let mockIsSubmitting: WritableSignal<boolean>;
+  let mockTaskId: WritableSignal<string | null>;
+  let mockQuizCompleted: WritableSignal<boolean>;
+
+  beforeEach(async () => {
+    // Initialize all signals
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.spec.ts]:
```suggestion
```

File: src/app/features/written-response/written-response.spec.ts [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+
+    mockActivatedRoute = {
+      snapshot: {
+        paramMap: {
+          get: jest.fn().mockReturnValue('task-123'),
+        },
+      },
+    };
+
+    // Setup selectSignal to return appropriate signals
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.spec.ts]:
```suggestion
```

File: src/app/features/written-response/written-response.spec.ts [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+      if (selectorString.includes('Error')) return mockError;
+      if (selectorString.includes('ExpectedDuration')) return mockExpectedDuration;
+      if (selectorString.includes('IsSubmitting')) return mockIsSubmitting;
+      if (selectorString.includes('TaskId')) return mockTaskId;
+      if (selectorString.includes('QuizCompleted')) return mockQuizCompleted;
+
+      return mockTaskTitle;
+    });
+
+    // Setup select to return observable
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.spec.ts]:
```suggestion
```

File: src/app/features/written-response/written-response.ts [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+  ) {}
+
+  ngOnInit(): void {
+    const routeTaskId = this.route.snapshot.paramMap.get('id');
+
+    if (routeTaskId) {
+      this.store.dispatch(WrittenResponseActions.loadWrittenResponseTask({ taskId: routeTaskId }));
+
+      this.taskSubscription = this.store
+        .select(selectWrittenResponseTask)
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: use a select signal here or make sure you unsubscribe

File: src/app/features/written-response/written-response.ts [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+        .pipe(filter((task) => task !== null))
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: use a select signal here or make sure you unsubscribe

File: src/app/features/written-response/store/written-response.effects.ts (hunk 16 above)
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/store/written-response.effects.ts]: let us destructure here

**Comment by Proc3ssa** [comment]
I have made the requested changes.

File: src/app/features/written-response/written-response.spec.ts [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+  };
+  let mockActivatedRoute: {
+    snapshot: {
+      paramMap: {
+        get: jest.Mock;
+      };
+    };
+  };
+
+  // Mock signals
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.spec.ts]:
```suggestion
```

File: src/app/features/written-response/written-response.ts [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+    private router: Router,
+  ) {
+    effect(() => {
+      const task = this.task();
+      if (task && !this.intervalId && !this.timerStarted) {
+        this.timerStarted = true;
+        const duration = task.estimatedDurationInMinutes || 10;
+        const totalSeconds = duration * 60;
+        this.startTimer(totalSeconds);
+      }
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/features/written-response/written-response.ts]: Using effect for this functionality is not the best approach. Check the documentation o when to use Effects and find a better approach

**Review by Kwabena-Owusu-Achiaw — APPROVED** [review]
